future/atomicTransactionComposer.go

```diff
@@ -0,0 +1,681 @@
+	signedTxs [][]byte
+
+	/** Txids of the transactions in this group. Populated after invocation to GatherSignatures() and used
+	in Execute() to gather transaction information */
+	txids map[int]string
+}
+
+func MakeAtomicTransactionComposer() AtomicTransactionComposer {
+	return AtomicTransactionComposer{
+		status:       BUILDING,
+		transactions: []TransactionWithSigner{},
+		methodCalls:  map[int]Method{},
+		signedTxs:    [][]byte{},
+		txids:        map[int]string{},
+	}
+}
+
+/**
+* Get the status of this composer's transaction group.
+ */
+func (atc *AtomicTransactionComposer) GetStatus() AtomicTransactionComposerStatus {
+	return atc.status
+}
+
+/**
+* Get the number of transactions currently in this atomic group.
+ */
+func (atc *AtomicTransactionComposer) Count() int {
+	return len(atc.transactions)
+}
+
+/**
+* Create a new composer with the same underlying transactions. The new composer's status will be
+* BUILDING, so additional transactions may be added to it.
+ */
+func (atc *AtomicTransactionComposer) Clone() AtomicTransactionComposer {
+	newTxs := make([]TransactionWithSigner, len(atc.transactions))
+	copy(newTxs, atc.transactions)
+	for i := range newTxs {
+		newTxs[i].Txn.Group = types.Digest{}
+	}
+
+	newMethodCalls := make(map[int]Method, len(atc.methodCalls))
+	for k, v := range atc.methodCalls {
+		newMethodCalls[k] = v
+	}
+
+	return AtomicTransactionComposer{
+		status:       BUILDING,
+		transactions: newTxs,
+		methodCalls:  newMethodCalls,
+		signedTxs:    [][]byte{},
+		txids:        map[int]string{},
+	}
+}
+
+func (atc *AtomicTransactionComposer) validateTransaction(txn types.Transaction) error {
+	emtpyGroup := types.Digest{}
+	if txn.Group != emtpyGroup {
+		return fmt.Errorf("expected empty group id")
+	}
+
+	return nil
+}
+
+/**
+* Add a transaction to this atomic group.
+*
+* An error will be thrown if the composer's status is not BUILDING, or if adding this transaction
+* causes the current group to exceed MAX_GROUP_SIZE.
+ */
+func (atc *AtomicTransactionComposer) AddTransaction(txnAndSigner TransactionWithSigner) error {
+	if atc.status != BUILDING {
+		return errors.New("status must be BUILDING in order to add tranactions")
+	}
+
+	if atc.Count() == MAX_GROUP_SIZE {
+		return fmt.Errorf("reached max group size: %d", MAX_GROUP_SIZE)
+	}
+
+	err := atc.validateTransaction(txnAndSigner.Txn)
+	if err != nil {
+		return err
+	}
+
+	atc.transactions = append(atc.transactions, txnAndSigner)
+	return nil
+}
+
+/**
+* Add a smart contract method call to this atomic group.
+*
+* An error will be thrown if the composer's status is not BUILDING, if adding this transaction
+* causes the current group to exceed MAX_GROUP_SIZE, or if the provided arguments are invalid
+* for the given method.
+ */
+func (atc *AtomicTransactionComposer) AddMethodCall(
+	params AddMethoCallParams,
+	/** A transaction signer that can authorize this application call from sender */
+	signer TransactionSigner) error {
+
```

Comment:
It seemed like a logical separation to me since all the other params are part of the appCall tx while the signer exists outside of the tx. I don't feel too strongly so I don't mind changing it.